### ISSUE TITLE
toplev: Only showing Frontend_Bound without -v

### ISSUE THREAD

**Issue by ddcc** [andikleen/pmu-tools#262]
It's my first time using toplev, so I may be doing something wrong. On a ~18 sec benchmark, regardless of whether I run at level 1, 2, or 3, I'm not getting much of a breakdown unless I pass the `-v` flag, as seen below. This seems unusual, so am I missing kernel support or hardware event counters?

This is a Coffee Lake processor on a Ubuntu 19.10 system running kernel 5.3.0, with a manually compiled perf for better unwinding, but I don't see any difference with the regular distro perf. I was previously having issues with divide by zeroes and missing events even at level 1, until I disabled hyperthreading.

```
$ PERF=~/linux-5.3.0/tools/perf/perf ~/pmu-tools/toplev.py -l1 --single-thread -v ./run-orig.sh
Using level 1.
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xd,umask=0x1/,cpu/event=0xe,umask=0x1/,cycles}' ./run-orig.sh
# 4.0-full on Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
FE      Frontend_Bound   % Slots            32.04   <==
        Sampling events:  frontend_retired.latency_ge_8:pp
BAD     Bad_Speculation  % Slots            8.31 ?
BE      Backend_Bound    % Slots            5.01 ?
RET     Retiring         % Slots            54.64 ?
        Sampling events:  uops_retired.retire_slots
MUX                      %                  100.00
```

```
$ PERF=~/linux-5.3.0/tools/perf/perf ~/pmu-tools/toplev.py -l2 --single-thread -v ./run-orig.sh
Using level 2.
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x79,umask=0x30/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc3,umask=0x1,edge=1,cmask=1/,cpu/event=0xc5,umask=0x0/,cpu/event=0xd,umask=0x1/,cpu/event=0xe,umask=0x1/,cycles},{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400106/,cpu/event=0xc6,umask=0x1,frontend=0x400206/,cpu/event=0xe,umask=0x1/,cycles,instructions},{cpu/event=0xa3,umask=0x14,cmask=20/,cpu/event=0xa6,umask=0x1/,cpu/event=0xa6,umask=0x2/,cpu/event=0xa6,umask=0x4/,cpu/event=0xa6,umask=0x40/,cpu/event=0xc2,umask=0x2/,cycles}' ./run-orig.sh
1 events not supported, 7 events not counted
# 4.0-full on Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
FE             Frontend_Bound                      % Slots                  31.90   <==
        Sampling events:  frontend_retired.latency_ge_8:pp
BAD            Bad_Speculation                     % Slots                   8.35 ?
BE             Backend_Bound                       % Slots                   4.87 ?
RET            Retiring                            % Slots                  54.89 ?
        Sampling events:  uops_retired.retire_slots
FE             Frontend_Bound.Frontend_Latency     % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_16:pp frontend_retired.latency_ge_32:pp
FE             Frontend_Bound.Frontend_Bandwidth   % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_2_bubbles_ge_1:pp
BAD            Bad_Speculation.Branch_Mispredicts  % Slots                   8.05 ?
        Sampling events:  br_misp_retired.all_branches
BAD            Bad_Speculation.Machine_Clears      % Slots                   0.28 ?
        Sampling events:  machine_clears.count
BE/Mem         Backend_Bound.Memory_Bound          % Slots                   1.41 ?
BE/Core        Backend_Bound.Core_Bound            % Slots                   3.46 ?
RET            Retiring.Base                       % Slots                  51.86 ?
        Sampling events:  inst_retired.prec_dist:pp
RET            Retiring.Microcode_Sequencer        % Slots                   3.03 ?
        Sampling events:  idq.ms_uops
MUX                                                %                        33.33
        PerfMon Event Multiplexing accuracy indicator
warning: 2 division by zero errors:Frontend_Bandwidth Frontend_Latency
```

```
$ PERF=~/linux-5.3.0/tools/perf/perf ~/pmu-tools/toplev.py -l3 --single-thread -v ./run-orig.sh
Using level 3.
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x79,umask=0x30/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc3,umask=0x1,edge=1,cmask=1/,cpu/event=0xc5,umask=0x0/,cpu/event=0xd,umask=0x1/,cpu/event=0xe,umask=0x1/,cycles},{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400106/,cpu/event=0xc6,umask=0x1,frontend=0x400206/,cpu/event=0xe,umask=0x1/,cycles,instructions},{cpu/event=0xa3,umask=0x14,cmask=20/,cpu/event=0xa6,umask=0x1/,cpu/event=0xa6,umask=0x2/,cpu/event=0xa6,umask=0x4/,cpu/event=0xa6,umask=0x40/,cpu/event=0xc2,umask=0x2/,cycles},{cpu/event=0x14,umask=0x1,cmask=1/,cpu/event=0x79,umask=0x30,edge=1,cmask=1/,cpu/event=0x80,umask=0x4,cmask=1,edge=1/,cpu/event=0x80,umask=0x4/,cpu/event=0x83,umask=0x4/,cpu/event=0x87,umask=0x1/,cpu/event=0xa6,umask=0x40/,cpu/event=0xab,umask=0x2/,cycles},{cpu/event=0x79,umask=0x18,cmask=1/,cpu/event=0x79,umask=0x18,cmask=4/,cpu/event=0x79,umask=0x24,cmask=1/,cpu/event=0x79,umask=0x24,cmask=4/,cpu/event=0xa8,umask=0x1,cmask=1/,cpu/event=0xa8,umask=0x1,cmask=4/,cpu/event=0xd,umask=0x80/,cpu/event=0xe6,umask=0x1/,cycles},{cpu/event=0xa3,umask=0x14,cmask=20/,cpu/event=0xa3,umask=0x5,cmask=5/,cpu/event=0xa3,umask=0x6,cmask=6/,cpu/event=0xa3,umask=0xc,cmask=12/,cpu/event=0xc1,umask=0x3f/,cpu/event=0xca,umask=0x1e,cmask=1/,cycles},{cpu/event$0x14,umask=0x1,cmask=1/,cpu/event=0xa6,umask=0x1/,cpu/event=0xa6,umask=0x2/,cpu/event=0xa6,umask=0x4/,cpu/event=0xc2$umask=0x2/,cycles},{cpu/event=0x48,umask=0x2,cmask=1/,cpu/event=0xa3,umask=0x5,cmask=5/,cpu/event=0xa3,umask=0x6,cma$k=6/,cpu/event=0xa3,umask=0xc,cmask=12/,cpu/event=0xd1,umask=0x2/,cpu/event=0xd1,umask=0x40/,cpu/event=0xd1,umask=0x$/,cycles},{cpu/event=0xb1,umask=0x1/,cpu/event=0xb1,umask=0x10/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc7,umask=0x1/,$pu/event=0xc7,umask=0x10/,cpu/event=0xc7,umask=0x2/,cpu/event=0xc7,umask=0x4/,cpu/event=0xc7,umask=0x8/},cpu/event=0$c7,umask=0x20/' ./run-orig.sh
1 events not supported, 23 events not counted
# 4.0-full on Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz                                                               FE             Frontend_Bound                                   % Slots                  31.70   <==
        Sampling events:  frontend_retired.latency_ge_8:pp
BAD            Bad_Speculation                                  % Slots                   8.07 ?
BE             Backend_Bound                                    % Slots                   5.03 ?
RET            Retiring                                         % Slots                  55.15 ?
        Sampling events:  uops_retired.retire_slots                                                                  
FE             Frontend_Bound.Frontend_Latency                  % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_16:pp frontend_retired.latency_ge_32:pp
FE             Frontend_Bound.Frontend_Bandwidth                % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_2_bubbles_ge_1:pp
BAD            Bad_Speculation.Branch_Mispredicts               % Slots                   7.85 ?
        Sampling events:  br_misp_retired.all_branches
BAD            Bad_Speculation.Machine_Clears                   % Slots                   0.29 ?

        Sampling events:  machine_clears.count
BE/Mem         Backend_Bound.Memory_Bound                       % Slots                   1.44 ?
BE/Core        Backend_Bound.Core_Bound                         % Slots                   3.59 ?
RET            Retiring.Base                                    % Slots                  52.09 ?
        Sampling events:  inst_retired.prec_dist:pp
RET            Retiring.Microcode_Sequencer                     % Slots                   3.06 ?
        Sampling events:  idq.ms_uops
FE             Frontend_Bound.Frontend_Latency.ICache_Misses    % Clocks                  0.00 ?
        Sampling events:  frontend_retired.l2_miss:pp frontend_retired.l1i_miss:pp
FE             Frontend_Bound.Frontend_Latency.ITLB_Misses      % Clocks                  0.00 ?
        Sampling events:  frontend_retired.stlb_miss:pp frontend_retired.itlb_miss:pp
FE             Frontend_Bound.Frontend_Latency.Branch_Resteers  % Clocks_est              0.00 ?
        Sampling events:  br_misp_retired.all_branches
FE             Frontend_Bound.Frontend_Latency.DSB_Switches     % Clocks                  0.00 ?
        Sampling events:  frontend_retired.dsb_miss:pp
FE             Frontend_Bound.Frontend_Latency.LCP              % Clocks                  0.00 ?
FE             Frontend_Bound.Frontend_Latency.MS_Switches      % Clocks                  0.00 ?
        Sampling events:  idq.ms_switches
FE             Frontend_Bound.Frontend_Bandwidth.MITE           % Core_Clocks             0.00 ?
        Sampling events:  frontend_retired.dsb_miss:pp
FE             Frontend_Bound.Frontend_Bandwidth.DSB            % Core_Clocks             0.00 ?
FE             Frontend_Bound.Frontend_Bandwidth.LSD            % Core_Clocks             0.00 ?
BE/Mem         Backend_Bound.Memory_Bound.L1_Bound              % Stalls                  5.52 ?
        Sampling events:  mem_load_retired.l1_hit:pp mem_load_retired.fb_hit:pp
BE/Mem         Backend_Bound.Memory_Bound.L2_Bound              % Stalls                  0.66 ?
        Sampling events:  mem_load_retired.l2_hit:pp
BE/Mem         Backend_Bound.Memory_Bound.L3_Bound              % Stalls                  1.42 ?
        Sampling events:  mem_load_retired.l3_hit:pp
BE/Mem         Backend_Bound.Memory_Bound.DRAM_Bound            % Stalls                  0.53 ?
        Sampling events:  mem_load_retired.l3_miss:pp
BE/Mem         Backend_Bound.Memory_Bound.Store_Bound           % Stalls                  0.00 ?
        Sampling events:  mem_inst_retired.all_stores:pp
BE/Core        Backend_Bound.Core_Bound.Divider                 % Clocks                  0.00 ?
        Sampling events:  arith.divider_active
BE/Core        Backend_Bound.Core_Bound.Ports_Utilization       % Clocks                 20.39 ?
RET            Retiring.Base.FP_Arith                           % Uops                    0.16 ?
RET            Retiring.Base.Other                              % Uops                   99.84 ?
RET            Retiring.Microcode_Sequencer.Assists             % Slots_est               0.00 ?
        Sampling events:  other_assists.any
MUX                                                             %                         0.00
        PerfMon Event Multiplexing accuracy indicator
warning: 13 division by zero errors:LSD MITE LCP Frontend_Bandwidth ICache_Misses DSB_Switches Divider Branch_Resteers Store_Bound MS_Switches ITLB_Misses DSB Frontend_Latency
```

**Comment by andikleen** [comment]
Your application is frontend bound.  None of the other nodes cross the
threshold. There can be cases where a top level node cannot be further
qualified. Part of it might be multiplexing problems, see the FAQ.

**Comment by ddcc** [comment]
Thanks for the response, by FAQ are you referring to the [manual](https://github.com/andikleen/pmu-tools/wiki/toplev-manual)?

When I rerun with `--no-multiplex`, I get essentially the same results, e.g. at level 2: 
```
Using level 2.
RUN #1 of 3
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x79,umask=0x30/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc3,umask=0x1,edge=1,cmask=1/,cpu/event=0xc5,umask=0x0/,cpu/event=0xd,umask=0x1/,cpu/event=0xe,umask=0x1/,cycles}' ./run-orig.sh
RUN #2 of 3
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400106/,cpu/event=0xc6,umask=0x1,frontend=0x400206/,cpu/event=0xe,umask=0x1/,cycles,instructions}' ./run-orig.sh
1 events not supported, 7 events not counted
RUN #3 of 3
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0xa3,umask=0x14,cmask=20/,cpu/event=0xa6,umask=0x1/,cpu/event=0xa6,umask=0x2/,cpu/event=0xa6,umask=0x4/,cpu/event=0xa6,umask=0x40/,cpu/event=0xc2,umask=0x2/,cycles}' ./run-orig.sh
# 4.0-full on Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
FE             Frontend_Bound                      % Slots                  32.16   <==
        Sampling events:  frontend_retired.latency_ge_8:pp
BAD            Bad_Speculation                     % Slots                   8.15 ?
BE             Backend_Bound                       % Slots                   4.34 ?
RET            Retiring                            % Slots                  54.98 ?
        Sampling events:  uops_retired.retire_slots
FE             Frontend_Bound.Frontend_Latency     % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_16:pp frontend_retired.latency_ge_32:pp
FE             Frontend_Bound.Frontend_Bandwidth   % Slots                   0.00 ?
        Sampling events:  frontend_retired.latency_ge_2_bubbles_ge_1:pp
BAD            Bad_Speculation.Branch_Mispredicts  % Slots                   7.82 ?
        Sampling events:  br_misp_retired.all_branches
BAD            Bad_Speculation.Machine_Clears      % Slots                   0.25 ?
        Sampling events:  machine_clears.count
BE/Mem         Backend_Bound.Memory_Bound          % Slots                   1.26 ?
BE/Core        Backend_Bound.Core_Bound            % Slots                   3.08 ?
RET            Retiring.Base                       % Slots                  51.92 ?
        Sampling events:  inst_retired.prec_dist:pp
RET            Retiring.Microcode_Sequencer        % Slots                   3.06 ?
        Sampling events:  idq.ms_uops
MUX                                                %                       100.00
        PerfMon Event Multiplexing accuracy indicator
warning: 2 division by zero errors:Frontend_Bandwidth Frontend_Latency
```

But, it looks like there's something weird going on with the counters from run 2:
```
~/linux-5.3.0/tools/perf/perf stat -x\; -e '{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400106/,cpu/event=0xc6,umask=0x1,frontend=0x400206/,cpu/event=0xe,umask=0x1/,cycles,instructions}' ./run-orig.sh
<not counted>;;cpu/event=0x9c,umask=0x1,cmask=4/;0;100.00;;
<not counted>;;cpu/event=0x9c,umask=0x1/;0;100.00;;
<not counted>;;cpu/event=0xc2,umask=0x2/;0;100.00;;
<not counted>;;cpu/event=0xc6,umask=0x1,frontend=0x400106/;0;100.00;;
<not supported>;;cpu/event=0xc6,umask=0x1,frontend=0x400206/;0;100.00;;
<not counted>;;cpu/event=0xe,umask=0x1/;0;100.00;;
<not counted>;;cycles;0;100.00;;
<not counted>;;instructions;0;100.00;;
```

More specifically, these two counters can't be run together in the same group: `cpu/event=0xc6,umask=0x1,frontend=0x400106/` and `cpu/event=0xc6,umask=0x1,frontend=0x400206/`. If I remove either and run them separately, then everything works fine:

```
perf stat -x\; -e '{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400106/,cpu/event=0xe,umask=0x1/,cycles,instructions}' ./run-orig.sh
12908823398;;cpu/event=0x9c,umask=0x1,cmask=4/;17683263963;100.00;;
112347415012;;cpu/event=0x9c,umask=0x1/;17683263963;100.00;;
194964671311;;cpu/event=0xc2,umask=0x2/;17683263963;100.00;;
5314904137;;cpu/event=0xc6,umask=0x1,frontend=0x400106/;17683263963;100.00;;
213358458322;;cpu/event=0xe,umask=0x1/;17683263963;100.00;;
88366683700;;cycles;17683263963;100.00;;
186441854574;;instructions;17683263963;100.00;2.11;insn per cycle
```

```
perf stat -x\; -e '{cpu/event=0x9c,umask=0x1,cmask=4/,cpu/event=0x9c,umask=0x1/,cpu/event=0xc2,umask=0x2/,cpu/event=0xc6,umask=0x1,frontend=0x400206/,cpu/event=0xe,umask=0x1/,cycles,instructions}' ./run-orig.sh
13430993435;;cpu/event=0x9c,umask=0x1,cmask=4/;17808261456;100.00;;
114546897568;;cpu/event=0x9c,umask=0x1/;17808261456;100.00;;
195005617312;;cpu/event=0xc2,umask=0x2/;17808261456;100.00;;
1184263589;;cpu/event=0xc6,umask=0x1,frontend=0x400206/;17808261456;100.00;;
213274944944;;cpu/event=0xe,umask=0x1/;17808261456;100.00;;
88979484325;;cycles;17808261456;100.00;;
186442304774;;instructions;17808261456;100.00;2.10;insn per cycle
```

For reference, this application is the SPEC2006 483.xalancbmk benchmark, which has a [published toplevel breakdown](http://www.cs.technion.ac.il/~erangi/TMA_using_Linux_perf__Ahmad_Yasin.pdf) that is significantly different, although my setup isn't exactly identicaly (Clang 10 instead of ICC 14, etc).

**Comment by andikleen** [comment]
With the FAQ I meant https://github.com/andikleen/pmu-tools/wiki/reasons-for-toplev-measuring-issues

The non scheduling group is a bug. Let make take a look at it.

Yes that sometimes higher nodes are not resolvable by lower nodes is a quirk of the thresholds. The thresholds are essentially heuristics and they don't always work out.

**Comment by ddcc** [comment]
Great, thanks for your help!

**Comment by andikleen** [comment]
The scheduling bug should be fixed with c87ed76d51523

**Comment by ddcc** [comment]
Thanks!
